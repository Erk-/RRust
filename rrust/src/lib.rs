#[macro_export]
macro_rules! rfn {
    ($name:ident, ($($param:ident: $party:ty),* $(,)?), $code:block) => {
        struct $name;

        impl $name {
            fn forward($($param:$party),*) {
                ::rrust::forward! {
                    $code
                };
            }
            fn backwards($($param:$party),*) {
                ::rrust::reverse! {
                    $code
                };
            }
        }
    }
}

#[macro_export]
macro_rules! rif {
    ($before:expr, $then:block, $else:block, $after:expr) => {
        if $before {
            ::rrust::forward! {
                $then
            };
            assert!($after);
        } else {
            ::rrust::forward! {
                $else
            };
            assert!(!($after));
        }
    };
    ($before:expr, $then:block, $after:expr) => {
        if $before {
            ::rrust::forward! {
                $then
            };
            assert!($after);
        }
    };
}

#[macro_export]
macro_rules! _reverse_rif {
    ($before:expr, $then:block, $else:block, $after:expr) => {
        if $after {
            ::rrust::reverse! {
                $then
            };
            assert!($before);
        } else {
            ::rrust::reverse! {
                $else
            };
            assert!(!($before));
        }
    };
    ($before:expr, $then:block, $after:expr) => {
        if $after {
            ::rrust::reverse! {
                $then
            };
            assert!($before);
        }
    };
}

#[macro_export]
macro_rules! rloop {
    ($from:expr, $do:block, $loop:block, $until:expr) => {
        assert!($from);
        ::rrust::forward! {
            $do
        };
        while !$until {
            ::rrust::forward! {
                $loop
            };
            assert!(!$from);
            ::rrust::forward! {
                $do
            };
        }
    };
    ($from:expr, $loop:block, $until:expr) => {
        assert!($from);
        while !$until {
            ::rrust::forward! {
                $loop
            };
            assert!(!$from);
        }
    };
}

#[macro_export]
macro_rules! _reverse_rloop {
    ($from:expr, $do:block, $loop:block, $until:expr) => {
        assert!($until);
        ::rrust::reverse! {
            $do;
        };
        while !$from {
            ::rrust::reverse! {
                $loop;
            };
            assert!(!$until);
            ::rrust::reverse! {
                $do;
            };
        }
    };
    ($from:expr, $loop:block, $until:expr) => {
        assert!($until);
        while !$from {
            ::rrust::reverse! {
                $loop
            };
            assert!(!$until);
        }
    };
}

#[doc(hidden)]
pub use rrust_macro::{forward, reverse};

#[macro_export]
macro_rules! delocal {
    ($name:ident, $e:expr) => {
        if $name != $e {
            panic!("Delocal failed {} != {}", $name, $e);
        }
        drop($name);
    };
}

