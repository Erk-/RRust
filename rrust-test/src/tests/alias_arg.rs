use rrust::rfn;

rfn!(AliasArg, (x1: &mut i32, x2: &mut i32), {
    *x1 += 1;
    *x2 += 1;
});

fn main() {
    let mut x = 0;

    AliasArg::forward(&mut x, &mut x);
    AliasArg::backwards(&mut x, &mut x);
}
