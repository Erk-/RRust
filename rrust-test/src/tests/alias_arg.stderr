error[E0499]: cannot borrow `x` as mutable more than once at a time
  --> src/tests/alias_arg.rs:11:31
   |
11 |     AliasArg::forward(&mut x, &mut x);
   |     ----------------- ------  ^^^^^^ second mutable borrow occurs here
   |     |                 |
   |     |                 first mutable borrow occurs here
   |     first borrow later used by call

error[E0499]: cannot borrow `x` as mutable more than once at a time
  --> src/tests/alias_arg.rs:12:33
   |
12 |     AliasArg::backwards(&mut x, &mut x);
   |     ------------------- ------  ^^^^^^ second mutable borrow occurs here
   |     |                   |
   |     |                   first mutable borrow occurs here
   |     first borrow later used by call
